ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,496 @@
+    modules with internal things.
+* Avoid cluttering the top-level `Temporalio` module with many small classes, but some are ok as we try to avoid a
+  `Temporalio::Common` module
+  * So `Temporalio::RetryPolicy` is preferred over `Temporalio::Common::RetryPolicy`.
+  * The search attribute collection is `Temporalio::SearchAttributes` and all things are under it, e.g.
+    `Temporalio::SearchAttributes::Key`.
+    * 💭 Why not `Temporalio::SearchAttribute` module with the collection underneath? The object has value on its own
+      so might as well make it a class.
+  * The base metric is `Temporalio::Metric`, and all metric things are under there, e.g. `Temporalio::Metric::Meter`.
+    * 💭 Why not `Temporalio::Metrics` module? The object has value on its own so might as well make it a class.
+   (e.g. `Temporalio::SearchAttribute::Key`)
+    and all metric things under `Temporalio::Metric` (e.g. `Temporalio::Metric::Meter`).
+  * 💭 Why no `Temporalio::Common`? Ruby expects to have common things where they make sense instead of under a `Common`
+    module.
+* No compatibility with any existing Ruby SDK
+  * Migration guides/docs may appear as needed.
+  * 💭 Why? Need a clean slate to develop a high-quality SDK from first principles.
+* Rust bridge uses [rb-sys](https://github.com/oxidize-rb/rb-sys) and [magnus](https://github.com/matsadler/magnus)
+  * 💭 Why not just `ruby.h` directly or just `rb-sys`? There are many things difficult to get right with data
+    conversion and we don't feel as if using either library is compromising  our ability to move later if needed.
+* Obvious general repo/project rules apply.
+  * Linting and auto-formatting (Rubocop check/autocorrect but have to customize Rubocop a lot because its defaults are
+    very dogmatic).
+  * Type checking (Steep).
+  * Deployed to RubyGems as separate gems for pure source, x64 linux, x64 darwin, x64 windows (i.e. mingw-ucrt), aarch64
+    linux (i.e. arm64), and arm64 darwin.
+  * CI against supported platforms on oldest and newest Ruby versions supported.
+  * MIT licensed.
+  * Full docs as README.
+  * No need for ruby.temporal.io, https://www.rubydoc.info/gems/temporalio is acceptable.
+  * Full integration test suite.
+
+### Basic Features
+
+By the very first version, the Ruby SDK should have feature parity with every other SDK (except in cases where we have
+decided other features in other SDKs will never be ported). Here's a non-exhaustive list of features we should include
+that are not discussed later:
+
+* Interceptors and an OpenTelemetry implementation
+* Custom metrics and optional metrics buffer for custom processing
+* Typed search attributes, memos, headers, etc
+* API key, RPC metadata, HTTP CONNECT proxy support, etc
+* Loggers with contextual information
+* Full error hierarchy with `Temporalio::Error` as the root
+* ❓ Are there any Ruby libraries so widely adopted they might as well be in the standard library and therefore deserve
```

Comment:
ActiveSupport is probably the big contender here, but TBH I don't know how much the Temporal would really benefit from having it available.

Optional support for things like `ActiveSupport::Duration` might be nice for things like `start_delay`, but that could be a pain to implement (& maintain) alongside normal `Numeric` arguments.

Reply:
:+1: We definitely plan to have samples showing integrations, but not sure we need an explicit integration for this. Whatever integrations we do have would be optional/additional because it is really important that we have as few dependencies as possible in the primary SDK because they our dependency constraints become users' transitive constraints. I believe we can have only 1 dependency (protobuf) at this time. `temporalio-opentelemetry` is the only additional gem I can think we'll have at first.